solidity/random-beacon/contracts/RandomBeacon.sol

```diff
@@ -780,6 +792,41 @@ contract RandomBeacon is Ownable {
         }
     }
 
+    // TODO: Documentation.
+    function notifyFailedHeartbeat(Heartbeat.FailureClaim calldata claim)
+        external
+    {
+        require(
+            keccak256(
+                abi.encodePacked(claim.groupPubKey, claim.groupMembers)
+            ) == groups.getGroupChecksum(claim.groupId),
+            "Passed group data are wrong"
+        );
```

Comment:
Those cases are not the same and looking for consistency here may not be a good idea. Personally, I prefer to do this validation directly in `RandomBeacon` contract and let the libraries do the processing on validated data. This way we don't tie the checksum stuff with the library and `RandomBeacon` is the only place where everything is integrated together. However, `submitRelayEntry` is an exception because we need to check the lifecycle of the relay entry. There is no sense to validate group data if relay entry is not in progress or timed out. We first need to make sure we can call `submitRelayEntry` and then check the group checksum because then we know `currentRequest.groupId` is a valid one and is not, for example `0` if entry is not in progress.